unobstructed_lanechange/src/unobstructed_lanechange.cpp

```diff
@@ -0,0 +1,614 @@
+        time_bound_points.reserve(time_boundary_exclusive_index);
+
+        if (time_boundary_exclusive_index == points.size())
+        {
+            time_bound_points.insert(time_bound_points.end(), points.begin(),
+                                    points.end());  // All points fit within time boundary
+        }
+        else
+        {
+            time_bound_points.insert(time_bound_points.end(), points.begin(),
+                                    points.begin() + time_boundary_exclusive_index - 1);  // Limit points by time boundary
+        }
+
+        return time_bound_points;
+
+    }
+    void UnobstructedLaneChangePlugin::splitPointSpeedPairs(const std::vector<PointSpeedPair>& points,
+                                                std::vector<lanelet::BasicPoint2d>* basic_points,
+                                                std::vector<double>* speeds)
+    {
+        basic_points->reserve(points.size());
+        speeds->reserve(points.size());
+
+        for (const auto& p : points)
+        {
+            basic_points->push_back(p.point);
+            speeds->push_back(p.speed);
+        }
+    }
+
+    int UnobstructedLaneChangePlugin::getNearestRouteIndex(lanelet::BasicLineString2d& points, const cav_msgs::VehicleState& state)
+    {
+        lanelet::BasicPoint2d veh_point(state.X_pos_global, state.Y_pos_global);
+                double min_distance = std::numeric_limits<double>::max();
+        int i = 0;
+        int best_index = 0;
+        for (const auto& p : points)
+        {
+            double distance = lanelet::geometry::distance2d(p,veh_point);
+            if (distance < min_distance)
+            {
+            best_index = i;
+            min_distance = distance;
+            }
+            i++;
+        }
+        return best_index;
+
+    }
+
+    int UnobstructedLaneChangePlugin::getNearestPointIndex(const std::vector<PointSpeedPair>& points,
+                                               const cav_msgs::VehicleState& state)
+    {
+        lanelet::BasicPoint2d veh_point(state.X_pos_global, state.Y_pos_global);
+        ROS_DEBUG_STREAM("veh_point: " << veh_point.x() << ", " << veh_point.y());
+        double min_distance = std::numeric_limits<double>::max();
+        int i = 0;
+        int best_index = 0;
+        for (const auto& p : points)
+        {
+            double distance = lanelet::geometry::distance2d(p.point, veh_point);
+            ROS_DEBUG_STREAM("distance: " << distance);
+            ROS_DEBUG_STREAM("p: " << p.point.x() << ", " << p.point.y());
+            if (distance < min_distance)
+            {
+            best_index = i;
+            min_distance = distance;
+            }
+            i++;
+        }
+
+        return best_index;
+    }
+
+    lanelet::BasicLineString2d  UnobstructedLaneChangePlugin::create_lanechange_route(lanelet::BasicPoint2d start, lanelet::ConstLanelet& start_lanelet, lanelet::BasicPoint2d end, lanelet::ConstLanelet& end_lanelet)
```

Comment:
Can we rename this to `create_lanechange_trajectory` or `create_lanechange_path`, since it doesn't actually create a route?